students/jsward/session09/mailroom_oo/cli_main.py

```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env python
+
+from donor_models import Donor, DonorCollection
+from operator import itemgetter
+
+############################
+# DONOR MANAGEMENT FUNCTIONS
+############################
+
+
+def add_donation():  # Tested
+    """
+    Adds a donation to the specified donor.  Prints a thank you letter to stdout.
+    :return: None
+    """
+    donor = find_donor()
+    while True:
+        donation_input = input("Input {}'s donation amount in dollars\n".format(donor.name))
+        try:
+            donor.add_donation(donation_input)
+            print(donor.thank_you_letter)
+            return
+        except ValueError:
+            print("{} is not a valid donation amount.".format(donation_input))
+            continue
+
+
+def add_donor():  # Tested
+    """
+    Adds a donor to the donor collection
+    :return: None
+    """
+    donor_valid = False
+    while not donor_valid:
+        try:
+            user_input = str(input("Provide the name to add to the donor list\n"))
+        except TypeError:
+            continue
+        try:
+            donor_collection.add_donor(user_input)
+            donor_valid = True
+        except ValueError:
+            print("Donor {} already exists.  Please provide a different donor name to add".format(user_input))
+            continue
+
+
+def list_donors():  # Tested
+    """
+    Prints a list of donor names to stdout.
+    :return: None
+    """
+    print(donor_collection.donor_list)
+
+
+def print_thank_you():  # Tested
+    """
+    Prompts for the name of a donor.  Prints a thank you letter for the donor's mmost recent donation.
+    :return: None
+    """
+    print(find_donor().thank_you_letter)
+
+
+def write_letters_to_disk():  # Tested
+    """
+    Writes a text file to disk for each donor in the donor collection.
+    Each file contains a thank you letter for the donor's most recent donation.
+    :return: None
+    """
+    for donor in donor_collection.donors:
+        with open('thank_you_letter_{}'.format(donor.name), 'w') as thank_you_file:
+            thank_you_file.write(donor.thank_you_letter)
+
+
+######################
+# NAVIGATION FUNCTIONS
+######################
+
+
+def donor_management():  # Tested
+    """
+    Calls a function based on the user's input in the program's donor menu.
+    :return: None
+    """
+    donor_menu_options = {'l': list_donors, 'a': add_donor, '$': add_donation, 't': print_thank_you,
+                          'g': write_letters_to_disk}
+    user_input = menu_donor()
+    donor_menu_options.get(user_input)()
+
+
+def find_donor():  # Tested
+    """
+    Prompts user for a donor name.  Locates the corresponding Donor object in the DonorCollection
+    :return: Donor object
+    """
+    while True:
+        user_input = str(input("Provide the name of the donor.\n"))
+        for donor in donor_collection.donors:
```

Comment:
It would have been good to implement a function in your donor_collection object to do this search for you, rather than have the cli_main handle this logic